Document/0x06f-Testing-Local-Authentication.md

```diff
@@ -9,26 +9,196 @@ As a result only requirement "4.6	Biometric authentication, if any, is not event
-
--- TODO [Describe how to assess this given either the source code or installer package (APK/IPA/etc.), but without running the app. Tailor this to the general situation (e.g., in some situations, having the decompiled classes is just as good as having the original source, in others it might make a bigger difference). If required, include a subsection about how to test with or without the original sources.] --
-
--- TODO [Confirm purpose of remark "Use the &lt;sup&gt; tag to reference external sources, e.g. Meyer's recipe for tomato soup<sup>[1]</sup>."] --
+Third-party apps have two ways to incorporate system-provided Touch ID authentication:
+* `LocalAuthentication.framework`<sup>[3]</sup> can be used to ask user to authenticate via Touch ID. The app can’t access the any data associated with the enrolled fingerprint and is notified only whether authentication was successful or not.
+* `Security.framwork`<sup>[4]</sup> gives access to Keychain Services, whose Access Control mechanism allows to protect Keychain items with Touch ID. This framework has C API, but there are dozens of 3rd party wrappers who makes work with Keychain as simple as with NSUserDefaults<sup>[5]</sup>. `Security.framework` is lower level API than  `LocalAuthentication.framework`, and is not always the most convenient option to use.
 
--- TODO [Add content for "Testing Biometric Authentication" with source code] --
+#### Static Analysis
+#### With Source Code
+##### LocalAuthentication.framework
+From developer's point of view, work with `LocalAuthentication.framework` is pretty straightforward: create instance of `LAContext`, ensure that OS and device support biometric authentication policy, evaluate policy with completion handler and title string explaining to user why she is requested to pass authentication right now.
+`LAPolicy` has two options:
+	* `deviceOwnerAuthentication`(Swift) or `LAPolicyDeviceOwnerAuthentication`(Objective-C) - using Touch ID or the device password
+	* `deviceOwnerAuthenticationWithBiometrics` (Swift) or `LAPolicyDeviceOwnerAuthenticationWithBiometrics`(Objective-C) - use Touch ID only
+
+Appropriate error will be returned to completion handler in following cases: authentication failed, user cancelled authentication, user have chosen fallback, system cancelled authentication, passcode is not set on device, TouchID is not available, TouchID is not enrolled.
+
+###### Swift
+```
+let context = LAContext()
+var error: NSError?
+
+guard context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) else {
+	// Could not evaluate policy; look at error and present an appropriate message to user
+}
+
+context.evaluatePolicy(.deviceOwnerAuthentication, localizedReason: "Please, pass authorisation to enter this area") { success, evaluationError in
+	guard success else {
+		// User did not authenticate successfully, look at evaluationError and take appropriate action
+	}
+
+	// User authenticated successfully, take appropriate action
+}
```

Comment:
Where did you get the code examples from? Please provide the link to it and put it in the reference section.